ratis-common/src/main/java/org/apache/ratis/util/BatchLogger.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ratis.util;
+
+import org.apache.ratis.thirdparty.com.google.common.cache.Cache;
+import org.apache.ratis.thirdparty.com.google.common.cache.CacheBuilder;
+import org.apache.ratis.thirdparty.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+public final class BatchLogger {
+
+  private BatchLogger() {
+  }
+
+  public final static class UniqueId {
+    private final List<Object> components;
+    private UniqueId(Object... components) {
+      this.components = Lists.newArrayList(components);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (!(obj instanceof UniqueId)) {
+        return false;
+      }
+
+      return components.equals(((UniqueId) obj).components);
+    }
+
+    @Override
+    public int hashCode() {
+      return components.hashCode();
+    }
+  }
+
+  public static UniqueId makeUniqueId(Object... components) {
+    return new UniqueId(components);
+  }
+
+  private final static class BatchedLogEntry {
+    private final Consumer<String> log;
+    private final AtomicInteger count;
+    private final Runnable logOp;
+    private final AtomicBoolean batchStarted;
+    private Timestamp startTime;
+
+    private BatchedLogEntry(Consumer<String> log, Runnable logOp) {
+      this.log = log;
+      this.logOp = logOp;
+      this.count = new AtomicInteger();
+      this.startTime = Timestamp.currentTime();
+      this.batchStarted = new AtomicBoolean(false);
+    }
+
+    private void increaseCount() {
+      this.count.incrementAndGet();
+    }
+
+    private synchronized void execute() {
+      log.accept(String.format("Received %s logs of following between %s and %s:",
+          count.get(), startTime, Timestamp.currentTime()));
+      logOp.run();
+      batchStarted.set(false);
+    }
+
+    private synchronized boolean tryStartBatch() {
```

Comment:
do we need synchronized here?

Reply:
I think it's necessary. The possible data race condition is:
One application thread call the `BatchLogger.warn` while the `SCHEDULER` thread decides that the batch duration is timeout and execute the batched log. Since both threads may access the `startTime` and `count`, we need synchronizations here to protect data integrity.

Reply:
That's is a good point!  Then, we may have dropped some messages in the current code in the following step:
1.  A thread has got an entry from `LOG_CACHE.computeIfAbsent(..)` but has not yet called `tryStartBatch(..)`.
2. The schedule thread removes the entry.

Then, the message in (1) will not be printed.  Let's use `startTime` to detect if an entry is already executed; see the other comments.